iscsi.go

```diff
@@ -0,0 +1,475 @@
+	"fmt"
+	"io"
+	"log"
+	"log/slog"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/avast/retry-go/v4"
+	_ "github.com/ianlancetaylor/cgosymbolizer"
+	gopointer "github.com/mattn/go-pointer"
+	"golang.org/x/sys/unix"
+)
+
+var defaultLogger = atomic.Pointer[slog.Logger]{}
+
+func init() {
+	defaultLogger.Store(slog.New(slog.NewTextHandler(io.Discard, &slog.HandlerOptions{
+		Level: slog.LevelError,
+	})))
+}
+
+func SetLogger(l *slog.Logger) {
+	defaultLogger.Store(l)
+}
+
+func logger() *slog.Logger {
+	return defaultLogger.Load()
+}
+
+type (
+	iscsiContext *C.struct_iscsi_context
+)
+
+type device struct {
+	Context      iscsiContext
+	targetName   string
+	targetPortal string
+	targetLun    int
+	details      ConnectionDetails
+}
+
+type ConnectionDetails struct {
+	InitiatorIQN string
+	TargetURL    string
+}
+
+func New(details ConnectionDetails) *device {
+	return &device{
+		details: details,
+	}
+}
+
+func (d *device) initializeContext() {
+	if d.Context != nil {
+		_ = C.iscsi_destroy_context(d.Context)
+		d.Context = nil
+		d.targetLun = 0
+		d.targetName = ""
+		d.targetPortal = ""
+	}
+	iqnStr := C.CString(d.details.InitiatorIQN)
+	defer C.free(unsafe.Pointer(iqnStr))
+	ctx := C.iscsi_create_context(iqnStr)
+	targetStr := C.CString(d.details.TargetURL)
+	defer C.free(unsafe.Pointer(targetStr))
+	url := C.iscsi_parse_full_url(ctx, targetStr)
+	_ = C.iscsi_set_targetname(ctx, &url.target[0])
+	defer C.iscsi_destroy_url(url)
+	d.Context = ctx
+	d.targetLun = int(url.lun)
+	d.targetName = C.GoString(&url.target[0])
+	d.targetPortal = C.GoString(&url.portal[0])
+	_ = C.iscsi_set_session_type(d.Context, C.ISCSI_SESSION_NORMAL)
+	_ = C.iscsi_set_header_digest(d.Context, C.ISCSI_HEADER_DIGEST_NONE_CRC32C)
+}
+
+func (d *device) Connect() error {
+	d.initializeContext()
+	return retry.Do(func() error {
+		portalStr := C.CString(d.targetPortal)
+		defer C.free(unsafe.Pointer(portalStr))
+		if retval := C.iscsi_full_connect_sync(d.Context, portalStr, C.int(d.targetLun)); retval != 0 {
+			errstr := C.iscsi_get_error(d.Context)
+			// reset the context before retrying.  it seems like some connection
+			// errors leave the context in an inconsistent state that makes it
+			// difficult to reuse
+			d.initializeContext()
+			return fmt.Errorf("iscsi_full_connect_sync: (%d) %s", retval, C.GoString(errstr))
+		}
+		return nil
+	}, retry.Attempts(20), retry.MaxDelay(500*time.Millisecond))
+}
+
+func (d *device) Reconnect() error {
+	if retval := C.iscsi_reconnect_sync(d.Context); retval != 0 {
+		if retval != 0 {
+			return fmt.Errorf("failed to reconnect with: %d", retval)
+		}
+	}
+	return nil
+}
+
+func (d *device) Disconnect() error {
+	defer C.iscsi_destroy_context(d.Context)
+	retval := C.iscsi_logout_sync(d.Context)
+	if retval != 0 {
+		return fmt.Errorf("failed to logout with: %d", retval)
+	}
+	return nil
+}
+
+type Capacity struct {
+	LBA       int
+	BlockSize int
+}
+
+func (d device) ReadCapacity10() (c Capacity, err error) {
```

Comment:
I understand the significance of 10/16 here, but I'm curious why we need to implement both? Is that just a best practice for consumers that may want to use the `10` functions? Just curious about this really.

Reply:
Oh I see we don't implement `Write10`, interesting.

Reply:
I implemented 10 first before I realized that it only supports devices up to 2TiB and so 16 would be needed in order to be able to implement an io.Reader over larger devices.  I'd like to eventually support the complete libiscsi api so I decided to just leave the 10.